Automattic/wp-calypso
# Plans 2023: Migrate plan-properties computation and other refactors

Fixes https://github.com/Automattic/wp-calypso/issues/79003, https://github.com/Automattic/wp-calypso/issues/79000
Partly addresses https://github.com/Automattic/wp-calypso/issues/79004, https://github.com/Automattic/wp-calypso/issues/79005 -- which can be deprioritized until package extraction

## Proposed Changes

- extracts the plan properties computation into respective hooks for putting together `GridPlan` concept. 
    - `GridPlan` is a more complete data structure with most data needed by the respective grid components i.e. components are now fed only the GridPlans they need to render (e.g. ComparisonGrid gets the full set of plans that features are built out from) and these include all Pricing and Features metadata.
- starts to surface more the splitting up of individual components - you will notice two instances of grid context provider, which is what the eventual split will be (with FeaturesGrid, ComparisonGrid, SpotlightPlan be enclosed within their own grid context)

A few more:

- More work/refactors on the hooks used for putting together GridPlan can help down the line (`use-grid-plan`, `use-plan-features-`). Nothing final, and the calls made in `plan-features-main` can potentially be reduced once everything public-facing for the plans package.
- Decouples the grid components from `client/lib/plans/features-list`  (some parts are intermediate until feature definitions migrate to calypso-products). So partly addresses https://github.com/Automattic/wp-calypso/issues/79004 to get us closer to a package
- Decouples the grid from local calypso state for pricing computation (also intermediate until site and plans data stores are built out more, although this looks like it will be a long way out endeavor). So also partly addresses https://github.com/Automattic/wp-calypso/issues/79005 to get us closer to a package

## Testing Instructions



Everything plans related:

@southp has written out a careful testing plan [down below](https://github.com/Automattic/wp-calypso/pull/79733#issuecomment-1651192466). For a more general sanity check:

- Test `/start` and tailored flows (e.g. `/start/hosting`, `/setup/newsletter`, `/setup/link-in-bio`)
- Test `/plans/[ paid, free]`
- Test `/plans` with WooExpress (grabbed from `/setup/wooexpress`) and Newsletter sites
- Confirm plan features render correctly in both features table and comparison grid (compare with production that all the checkmarks are correct)
- Confirm pricing, discounts, and notices render correctly (try `/plans/[paid]` with a paid site to show the pro-rated credits notice)

## Pre-merge Checklist
